lib/ui/painting/image_decoder.cc

```diff
@@ -57,30 +62,97 @@ static SkISize GetResizedDimensions(SkISize current_size,
+    FML_LOG(WARNING) << "Image is being upsized from "
+                     << image->dimensions().width() << "x"
+                     << image->dimensions().height() << " to "
+                     << resized_dimensions.width() << "x"
+                     << resized_dimensions.height()
+                     << ". Using cache(Height|Width) correctly?";
+    // TOOD(jeltsch): consider exiting here, there's no good reason to support
+    // upsampling in a "caching"-optimiztion context..
   }
 
-  TRACE_EVENT0("flutter", __FUNCTION__);
-  flow.Step(__FUNCTION__);
+  const bool aspect_ratio_changed =
+      std::abs(AspectRatio(resized_dimensions) -
+               AspectRatio(image->dimensions())) > 0.01;
+  if (aspect_ratio_changed) {
+    // This is probably a bug. If a user passes dimensions that change the
+    // aspect ratio in a "caching" context that's probably not working as
+    // intended and rather a signal that the API is hard to use.
+    FML_LOG(WARNING) << "Aspect ratio is being changed.";
+  }
 
-  const auto scaled_image_info = image->imageInfo().makeWH(
-      resized_dimensions.width(), resized_dimensions.height());
+  if (codec != nullptr) {
+    auto decode_size = codec->getScaledDimensions(
+        std::max(static_cast<double>(resized_dimensions.width()) /
+                     image->dimensions().width(),
+                 static_cast<double>(resized_dimensions.height()) /
+                     image->dimensions().height()));
+
+    // If the codec supports efficient sub-sampling during decode use the
+    // supported size instead.
+    // Note that this won't respect the exact user-requested size but instead
+    // choose a supported approximation.
+    if (decode_size != image->dimensions()) {
+      const auto scaled_image_info =
+          image->imageInfo().makeDimensions(decode_size);
+      SkBitmap scaled_bitmap;
+      if (!scaled_bitmap.tryAllocPixels(scaled_image_info)) {
+        FML_LOG(ERROR) << "Could not allocate bitmap when attempting to scale.";
+        return nullptr;
+      }
+
+      if (codec->getPixels(scaled_bitmap.pixmap()) ==
+          SkCodec::Result::kSuccess) {
+        // Marking this as immutable makes the MakeFromBitmap call share
+        // the pixels instead of copying.
+        scaled_bitmap.setImmutable();
+
+        auto decoded_image = SkImage::MakeFromBitmap(scaled_bitmap);
+        FML_DCHECK(decoded_image);
+        if (decoded_image) {
+          FML_LOG(ERROR)
+              << "Could not create a scaled image from a scaled bitmap.";
+        } else {
+          image = decoded_image;
+        }
+        // TODO(jeltsch): Depending on the intended semantics it might also be
+        // ok to return here. Concretely whether cacheHeight/cacheWidth
+        // semantically should be understood as sizing hints and/or whether
+        // changing aspect ratio or upsampling are really features we should
+        // support.
```

Comment:
We should file a bug to this effect and link it here.

Reply:
https://github.com/flutter/flutter/issues/48885